Feature/preview images

Quillraven opened the pull request:
Added a new `PreviewImageController` with two functions:

- `createPreviewImage`
- `getPreviewImage`

Example call for create:
```
curl --request POST \
  --url http://localhost:8080/api/previews/ \
  --header 'accept: text/event-stream' \
  --header 'content-type: multipart/form-data; boundary=---011000010111000001101001' \
  --form file= \
  --form Content-Type=image/png
```

Example call for get:
```
curl --request GET \
  --url http://localhost:8080/api/previews/5f4a65c642ce3b12422f66ab \
  --header 'accept: text/event-stream'
```

`createPreviewImage` stores a `GridFSFile` in mongodb which is linked to several chunks which is represented by a `GridFSResource`, I think.

Also, added a `ReactiveGridFSTemplate` bean to our Configuration. The template contains basic operations like `store`, `delete`, `update`, `find`, etc..

Finally, the ReactApp is now also showing an image at the top of the card that contains nearby toilets.

----

I am not 100% sure, if everything I did is a best practice. From my understanding I am downloading the bytes of an image on the server and send the bytes to the client. Maybe it is better to direct the `InputStream` to the client and let him get the image content?

Another thing is the `ResponseEntity`. I am not sure if a ResponseEntity should be used together with Webflux. Maybe it is better to remove the `ResponseEntity` and wrap the content(=body) directly inside a Mono/Flux?

Last, the current approach requires that the client sends the information what kind of image he is creating (jpg,png,...). Should it really be that way? Or should the server identify the `_contentType` that is passed to `GridFS`?

christian-baumann commented:
Seems quite complicated and I cannot see any advantages to the following straight forward solution:
E.g. download example:

Backend:
Just return the the image as ByteArray and set the correct value for `produces`:

```
@Controller
@RequestMapping("/")
class ImageController {
    @GetMapping("/image", produces = arrayOf(MediaType.IMAGE_JPEG_VALUE))
    @ResponseBody
    fun getImage(): ByteArray {
        val file = File("/home/cha/git/toilapp/frontend/public/toilet3.jpg")
        return file.readBytes()
    }
}
``` 

Frontend:
Just use the URL, e.g. if the Spring backends URL is localhost:8080 and the image is used within a Material-UI CardMedia element:

```
<CardMedia
          component="img"
          alt="Contemplative Reptile"
          height="140"
          image="http://localhost:8080/image"
          title="Contemplative Reptile"
        />
```

Quillraven commented:
@christian-baumann updated now several things:

- introduced services and moved the logic from the controllers to those new services
- create image now requires a toilet id to link the image to an existing toilet
- updated the backend example so that it shows the preview image per toilet (this is still the frontend from the master and it is not your improved version)
- added missing public folder to repo
- removed ResponseEntity entirely and always return now a Mono/Flux of a specific objecttype


There is an error case in the create image logic. If, for whatever reason, the toilet update call is not successful, then we create an image that is linked to nothing.
My idea was to use a transaction to avoid that meaning that either upload AND update are successful or a reroll should happen.
However, I was not able to make the `@Transactional` working. Seems like we'd need a replica set mongodb setup for that, and I did not have time to investigate it further.
There was once a `doInTransaction` function available but it is now deprecated and seems to be replaced by the transactionManager and the Transactional annotation.

Quillraven commented:
@christian-baumann added the `@Transactional` annotiation that should now help with the error case mentioned above. I did not test it yet but the error message that I got before when using this annotation is now gone. Therefore, I guess I correctly setup the mongo replica set environment.
I added the information to the Readme.md.
